syft/tensor.py

```diff
@@ -1,85 +1,169 @@
+    def dot(self, tensor):
+        """Returns inner product of two tensors"""
         if self.encrypted:
             return NotImplemented
 
-        arr_like = _ensure_ndarray(arr_like)
-        return self.data * arr_like
+        return syft.dot(self, tensor)
 
-    def __imul__(self, arr_like):
-        """Performs in place element-wise multiplication between two array like objects"""
+    def __matmul__(self, tensor):
+        """Performs matrix multiplication between two tensors"""
         if self.encrypted:
             return NotImplemented
 
-        arr_like = _ensure_ndarray(arr_like)
-        self.data = self.data * arr_like
-        return self.data
+        return syft.matmul(self, tensor)
 
-    def __truediv__(self, arr_like):
-        """Performs element-wise division between two array like objects"""
+    def __mul__(self, tensor):
+        """Performs element-wise multiplication between two tensors"""
         if self.encrypted:
             return NotImplemented
 
-        arr_like = _ensure_ndarray(arr_like)
-        return self.data / arr_like
+        tensor = _ensure_tensorbase(tensor)
+        return TensorBase(self.data * tensor.data)
 
-    def __itruediv__(self, arr_like):
-        """Performs in place element-wise subtraction between two array like objects"""
+    def __imul__(self, tensor):
+        """Performs in place element-wise multiplication between two tensors"""
         if self.encrypted:
             return NotImplemented
 
-        arr_like = _ensure_ndarray(arr_like)
-        self.data = self.data / arr_like
-        return self.data
+        tensor = _ensure_tensorbase(tensor)
+        self.data = self.data * tensor.data
+        return self
+
+    def __truediv__(self, tensor):
+        """Performs element-wise division between two tensors"""
+        if self.encrypted:
+            return NotImplemented
+
+        tensor = _ensure_tensorbase(tensor)
+        return TensorBase(self.data / tensor.data)
+
+    def __itruediv__(self, tensor):
+        """Performs in place element-wise subtraction between two tensors"""
+        if self.encrypted:
+            return NotImplemented
+
+        tensor = _ensure_tensorbase(tensor)
+        self.data = self.data / tensor.data
+        return self
+
+    def shape(self):
+        """Returns a tuple of input array dimensions."""
+        if self.encrypted:
+            return NotImplemented
+
+        return self.data.shape
+
+    def dim(self):
+        """Returns an integer of the number of dimensions of this tensor."""
+        if self.encrypted:
+            return NotImplemented
+
+        return self.data.ndim
+
+    def sum(self, dim=None):
+        """Returns the sum of all elements in the input array."""
+        if self.encrypted:
+            return NotImplemented
+
+        if dim is None:
+            return self.data.sum()
+        else:
+            return self.data.sum(axis=dim)
```

Comment:
I think there is no need for this `if` and `else`. Axis argument to `numpy.sum` can be `None`.